File: 0377-combination-sum-iv/0377-combination-sum-iv.java
class Solution {
    public int combinationSum4(int[] nums, int target) {
        final int[] dp = new int[target + 1];
        dp[0] = 1;
        for (int i = 1; i <= target; i++) {
            for (int k : nums) {
                if (i - k >= 0) {
                    dp[i] += dp[i - k];
                }
            }
        }
        return dp[target];
    }
}

File: 0377-combination-sum-iv/0377-combination-sum-iv.cpp
class Solution {
private:
    int dfs(vector<int>& nums, int target, unordered_map<int, int>& cache) {
        if (!target)
            return 1;
        if (!cache.count(target)) {
            cache[target] = 0;
            for (auto& num : nums)
                if (target - num >= 0)
                    cache[target] += dfs(nums, target - num, cache);
        }
        return cache[target];
    }

public:
    int combinationSum4(vector<int>& nums, int target) {
        unordered_map<int, int> cache;
        return dfs(nums, target, cache);
    }
};

// DFS + Memorization, time: O(n), space: O(n)